### ISSUE TITLE
Inheritance check

### ISSUE THREAD

**Issue by dioni21** [rodjek/puppet-lint#436]
Please reopen issue #304 

tl;dr:

    class foo {
      $a = 1
    }
    
    class bar inherits foo {
      notify { $a: }
    }

Shouldn't throw an error.

Reason: According to https://docs.puppetlabs.com/puppet/latest/reference/lang_classes.html#inheritance

> "The base class becomes the parent scope of the derived class, so that the new class receives a copy of all of the base class’s variables and resource defaults."

Otherwise, what would be the difference between inherits and include?

A command line option would work for me...

**Comment by dioni21** [comment]
Anybody?

**Comment by rnelson0** [comment]
The base class'es variables are still scoped with the class name, they are simply *available*. I.e. foo::a is copied into the scope of bar, but it is still known as foo::a. This is why when you see modules that inherit from a params class, they set default values to their parameters as foo::params::<var>.

I'm on a tablet so cannot provide better output and testing, but in that situation an examination of the catalog should show foo::a present, not bar::a. I will try and get that sample data and attach to this issue soon, to be sure.

**Comment by rnelson0** [comment]
Hrm, that's not what I expected! But probably not what most users would expect, either.
```
$ cat bar/manifests/init.pp
class bar inherits foo {
  notice("foo::a is ${foo::a}")
  notice("bar::a is ${bar::a}")
  notice("a is ${a}")
  notice("::a is ${::a}")
}

$ puppet apply --modulepath=. -e '$a = 5 include bar'
Notice: Scope(Class[Bar]): foo::a is 1
Notice: Scope(Class[Bar]): bar::a is 1
Notice: Scope(Class[Bar]): a is 1
Notice: Scope(Class[Bar]): ::a is 5

$ puppet-lint bar/manifests/init.pp
WARNING: class inherits across module namespaces on line 1
WARNING: top-scope variable being used without an explicit namespace on line 4
WARNING: class not documented on line 1
```
If you inherit from `foo::params`, then `$foo::params::a` is copied to `$bar::a`. Still, using `$a` is ambiguous as to whether it's top scope `$::a` or local scope `$::bar::a` since it was not assigned inside the class. If you make an assignment to `$a` in the class, the warning goes away because it is not ambiguous:
```
$ cat bar/manifests/init.pp
class bar inherits foo {
  $a = 72
  notice("foo::a is ${foo::a}")
  notice("bar::a is ${bar::a}")
  notice("a is ${a}")
  notice("::a is ${::a}")
}

$ puppet apply --modulepath=. -e '$a = 5 include bar'
Notice: Scope(Class[Bar]): foo::a is 1
Notice: Scope(Class[Bar]): bar::a is 72
Notice: Scope(Class[Bar]): a is 72
Notice: Scope(Class[Bar]): ::a is 5

$ puppet-lint bar/manifests/init.pp
WARNING: class inherits across module namespaces on line 1
WARNING: class not documented on line 1
```

Because of this, I see the warning as valid. Either refer to the full scoped variable that is inherited and assign it to a local var (as you often see with the params pattern) or refer to the inherited var in the local namespace using the full scope, or disable this lint check.

**Comment by rnelson0** [comment]
@dioni21 Let me know if you have any other questions before this is closed. Thanks!

**Comment by daenney** [comment]
Even though it might be valid compiler-wise this is also a linter, something that helps to enforce style. The style of inheriting from a class completely out of place (from `params` is the only exception) and then using the naked variables instead of the fully scoped ones is hard to trace down, especially if these things come from different modules.

Though the construct is valid, the community has long since agreed that this is undesirable. I think the warning is entirely warranted and it leads to more explicit/understandable code as a consequence.

**Comment by rnelson0** [comment]
I'm going to go ahead and mark this closed, but drop another note if you have questions and we can re-open if necessary. Thank you!
